templates/vsc/ts/notification-express/src/notification/middleware.ts

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Microsoft Corporation.Add commentMore actions
+// Licensed under the MIT license.
+
+import { Activity, ConversationReference } from "@microsoft/agents-activity";
+import { TurnContext, Middleware } from "@microsoft/agents-hosting";
+import { LocalConversationReferenceStore } from "./storage";
+
+/**
+ * @internal
+ */
+enum ActivityType {
+  CurrentBotInstalled,
+  CurrentBotMessaged,
+  CurrentBotUninstalled,
+  TeamDeleted,
+  TeamRestored,
+  Unknown,
+}
+
+export function getKey(reference: Partial<ConversationReference>): string {
+  // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+  return `_${reference.conversation?.tenantId}_${reference.conversation?.id}`;
+}
+
+export class NotificationMiddleware implements Middleware {
+  private readonly conversationReferenceStore: LocalConversationReferenceStore;
+
+  constructor(storage: LocalConversationReferenceStore) {
```

Comment:
minor: for production environment, users may need to use other storage other than a local store? In this case, we may need to define an interface for the store and allow users implement their own production ready store.
Not blocking but I feel it deserves to follow up. We also need to mention this in README after we figured out a solution.

Reply:
I agree that we should highlight in README because I met 2 cases that customers didn't realize the local file when deployed and the installation data was not persistent. We provide an example of Azure Table Storage in large scale notification sample.